nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,175 @@
+from contextlib import contextmanager
+from typing import Iterator, Optional
+import atexit
+import os
+import ptpython.repl
+import pty
+import subprocess
+import sys
+import tempfile
+import traceback
+
+import common
+
+from machine import Machine, NixStartScript
+
+
+class VLan ():
+    def __init__(self, nr: int, tmp_dir: str):
+        self.nr = nr
+        self.socket_dir = os.path.join(tmp_dir, f"vde{self.nr}.ctl")
+
+        self.process: Optional[subprocess.Popen] = None
+        self.pid: Optional[int] = None
+        self.fd: Optional[os.file] = None
+
+        # TODO: don't side-effect environment here
+        os.environ[f"QEMU_VDE_SOCKET_{self.nr}"] = self.socket_dir
+
+    def start(self):
+        pty_master, pty_slave = pty.openpty()
+
+        self.process = subprocess.Popen(
+            ["vde_switch", "-s", self.socket_dir, "--dirmode", "0700"],
+            stdin=pty_slave,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=False,
+        )
+        self.pid = self.process.pid
+        self.fd = os.fdopen(pty_master, "w")
+        self.fd.write("version\n")
+
+        # TODO: perl version checks if this can be read from
+        # an if not, dies. we could hang here forever. Fix it.
+        assert self.process.stdout is not None
+        self.process.stdout.readline()
+        if not os.path.exists(os.path.join(self.socket_dir, "ctl")):
+            raise Exception("cannot start vde_switch")
+
+    def release(self):
+        self.fd.close()
+        self.process.terminate()
+
+
+class Driver():
+    def __init__(
+        self,
+        logger,
+        vm_scripts,
+        keep_vm_state=False,
+        configure_python_repl=id,
+        machine_class=Machine,
+        vlan_class=VLan,
+    ):
+        """
+        Args:
+            - configure_python_repl: a function to configure ptpython.repl
+        """
+        self.log = logger.log_machinestate
+        self.configure_python_repl = configure_python_repl
+
+        tmp_dir = os.environ.get("TMPDIR", tempfile.gettempdir())
+        os.makedirs(tmp_dir, mode=0o700, exist_ok=True)
+
+        self.vlans = [
+            self.log(f"starting VDE switch for network {nr}") and
+            vlan_class(nr, tmp_dir) for nr in
+            list(dict.fromkeys(os.environ.get("VLANS", "").split()))
+        ]
+
+        def cmd(scripts):
+            for s in scripts:
+                yield NixStartScript(s)
+
+        self.machines = [
+            self.log(f"creating VM '{cmd.machine_name}'") and
```

Comment:
Maybe just move this log inside the `__init__` of the machine?

Reply:
It's still the machine controller who's logging here, there are actually other instances of that pattern and they are still within machine. I'll have to clean them up, too.

Reply:
Thought about it, definitly it wold be awkward if those already had the `[machine MSC]` prefix. Same with cvlan creation so I'll leave them here. Is a bit awkward but doesn't hurt and I wanted to keep it compact so that this driver init does not come along more complex than it actually is.